### PR TITLE
refactor RangeArgument

### PR DESCRIPTION
**Pull request by oli-obk**
implements #30877
supersedes #32681

r? @alexcrichton

### CODE REVIEW

**Review by durka — APPROVED** [review]

**Comment by brson** [comment]
In @sfackler's [last comment](https://github.com/rust-lang/rust/issues/30877#issuecomment-245174030). The `Bound` variants were called `Inclusive` and `Exclusive` whereas in this patch they are `Included` and `Excluded`. Is it appropriate to change them in this patch?

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -1349,9 +1350,16 @@ impl String {
...
+        let start = match range.lower() {
+            Bound::Included(&start) => start,
+            Bound::Excluded(&start) => start + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.upper() {
+            Bound::Included(&end) => end + 1,
+            Bound::Excluded(&end) => end,
+            Bound::Unbounded => self.len(),
+        };
```
Comment (brson) [inline src/libcollections/string.rs]: What is the contract wrt to wrapping of the bounds arithmetic here?

File: src/libcollections/vec.rs [hunk 2]
```diff
@@ -1060,8 +1061,16 @@ impl<T> Vec<T> {
...
+        let start = match range.lower() {
+            Bound::Included(&start) => start,
+            Bound::Excluded(&start) => start + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.upper() {
+            Bound::Included(&end) => end + 1,
+            Bound::Excluded(&end) => end,
+            Bound::Unbounded => len,
+        };
```
Comment (brson) [inline src/libcollections/vec.rs]: Ditto re wrapping.

**Comment by alexcrichton** [comment]
Oops sorry I think this accidentally fell out of my inbox, but thanks for the PR @oli-obk! I think that this is covered currently by https://github.com/rust-lang/rust/pull/38610 though, right?

File: src/libcollections/string.rs (hunk 1 above)
Comment (oli-obk) [inline src/libcollections/string.rs]: No issue, containers can't be bigger than isize::max afaik

**Comment by oli-obk** [comment]
Indeed, it is covered by that, although the bound methods are named differently.

**Pull request closed without merging**